src/lower.c

```diff
@@ -6283,27 +6286,27 @@ static void lower_stmt(tree_t stmt, loop_stack_t *loops)
 
    // TODO: Store differential path, compared to current block, not whole path
    //       to reduce size of coverage file.
-   ident_t hier = vcode_unit_name();
-   if (COV_ENABLED(OPT_COVER)) {
-      if (cover_is_stmt(stmt)) {
-
-         // Loops create scope, and statements are implicitly numbered within
-         // each scope. Correct hierarchical path must take into account loop
-         // name.
-         loop_stack_t *curr_loop = loops;
-         ident_t loop_prefix = NULL;
-         while (curr_loop) {
-            loop_prefix = ident_prefix(curr_loop->name, loop_prefix, '.');
-            curr_loop = curr_loop->up;
-         }
-         if (loop_prefix)
-            hier = ident_prefix(hier, loop_prefix, '.');
-         hier = ident_prefix(hier, tree_ident(stmt), '.');
+   if (COV_ENABLED(OPT_COVER) && cover_is_stmt(stmt)) {
+      ident_t label = tree_ident(stmt);
 
-         if (opt_get_int(OPT_COVER_STMT)) {
-            int32_t tag = cover_add_tag(stmt, hier, cover_tags, TAG_STMT, 0)->tag;
-            emit_cover_stmt(tag);
-         }
+      // Loops create scope, and statements are implicitly numbered within
+      // each scope. Correct hierarchical path must take into account loop
+      // name.
+      ident_t hier = vcode_unit_name();
+      loop_stack_t *curr_loop = loops;
+      ident_t loop_prefix = NULL;
+      while (curr_loop) {
+         loop_prefix = ident_prefix(curr_loop->name, loop_prefix, '.');
+         curr_loop = curr_loop->up;
+      }
+      if (loop_prefix)
+         hier = ident_prefix(hier, loop_prefix, '.');
+      hier = ident_prefix(hier, label, '.');
+      /// XXX: unused, redundant now?
```

Comment:
I am not sure.

What happends in case of nested loops? The original approach added "cover scope"
with each "language scope" (since also for loop which does not generate branch cover tag, creates scope).

Consider following example:

```
my_proc: process
        procedure my_wait(
            signal clk_sys : in std_logic
        ) is begin
            wait until rising_edge(clk_sys);
        end procedure;
    begin
        report "HALOO";
        for i in 0 to 10 loop         -- _L0
            for j in 1 to 5 loop       -- _L1
                my_wait(clk_sys);
            end loop;
        end loop;
        wait;
    end process;
```

The original approach gives cover tag path:

```
SPECT_TB.TB_TOP.MY_PROC._L0._L1._S0
```

but the new-one gives:

```
SPECT_TB.TB_TOP.MY_PROC._S0
```

This is basically the same discussion as previous comment. The ultimate question is:

For coverage, what objects should cause nesting in name of coverage tag hierarchy ?

IMHO, it should be everything, so:
 - Process
 - Any statement which might have sub-statements (if, case, for, while, etc...)
 - Branch name for all "if conditions" and "case assocs".
 
I would not use loop names to nest hierarchy, since according to LRM, they
increment linearly within process, not restart with each scope. 


Note that by that I don't mean "design hierarchy", that is still distinguished by
"COVER_TAG_HIER", so that report generation can recognize when to create
new file.

IMHO, this would be simplest and most straightforward for user.

What do you think?

Reply:
To be honest I don't really have a strong opinion either way. I think you should be able to get the effect you want by adding more cover push/pop calls? Maybe try modifying the code so you're happy with the scopes it generates, and then I'll take a look. TBH at the moment I'm mostly concerned with making sure the interface to the rest of the program is relatively simple and maintainable.

Reply:
I understand that. I will go through the push/pop and build up the coverage scopes hierarchy directly in the "lower". Then I can even completely remove the implicit labelling of statements in the "parse", and keep only
process and loop labeling as defined in LRM.